lib/netext/resolver.go

```diff
@@ -0,0 +1,140 @@
+ *
+ */
+
+package netext
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+)
+
+// MultiResolver returns all IP addresses for the given host.
+type MultiResolver func(host string) ([]net.IP, error)
+
+// Resolver is an interface that returns DNS information about a given host.
+type Resolver interface {
+	LookupIP(host string) (net.IP, error)
+}
+
+type resolver struct {
+	resolve    MultiResolver
+	strategy   lib.DNSStrategy
+	rrm        *sync.Mutex
+	rand       *rand.Rand
+	roundRobin map[string]uint8
+}
+
+type cacheRecord struct {
+	ips        []net.IP
+	lastLookup time.Time
+}
+
+type cacheResolver struct {
+	resolver
+	ttl   time.Duration
+	cm    *sync.Mutex
+	cache map[string]cacheRecord
+}
+
+// NewResolver returns a new DNS resolver. If ttl is not 0, responses
+// will be cached per host for the specified period. The IP returned from
+// LookupIP() will be selected based on the given strategy.
+func NewResolver(actRes MultiResolver, ttl time.Duration, strategy lib.DNSStrategy) Resolver {
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+	res := resolver{
+		resolve:    actRes,
+		strategy:   strategy,
+		rrm:        &sync.Mutex{},
+		rand:       r,
+		roundRobin: make(map[string]uint8),
+	}
+	if ttl == 0 {
+		return &res
+	}
+	return &cacheResolver{
+		resolver: res,
+		ttl:      ttl,
+		cm:       &sync.Mutex{},
+		cache:    make(map[string]cacheRecord),
+	}
+}
+
+// LookupIP returns a single IP resolved for host, selected by the
+// configured strategy.
+func (r *resolver) LookupIP(host string) (net.IP, error) {
+	ips, err := r.resolve(host)
+	if err != nil {
+		return nil, err
+	}
+	return r.selectOne(host, ips), nil
+}
+
+// LookupIP returns a single IP resolved for host, selected by the configured
+// strategy. Results are cached per host and will be refreshed if the last
+// lookup time exceeds the configured TTL (not the TTL returned in the DNS
+// record).
+func (r *cacheResolver) LookupIP(host string) (net.IP, error) {
+	r.cm.Lock()
+
+	var ips []net.IP
+	// TODO: Invalidate? When?
+	if d, ok := r.cache[host]; ok && time.Now().Before(d.lastLookup.Add(r.ttl)) {
+		ips = r.cache[host].ips
+	} else {
+		r.cm.Unlock() // The lookup could take some time, so unlock momentarily.
+		var err error
+		ips, err = r.resolve(host)
+		if err != nil {
+			return nil, err
+		}
+		r.cm.Lock()
+		r.cache[host] = cacheRecord{ips: ips, lastLookup: time.Now()}
+	}
+
+	r.cm.Unlock()
+	return r.selectOne(host, ips), nil
+}
+
+func (r *resolver) selectOne(host string, ips []net.IP) net.IP {
+	if len(ips) == 0 {
+		return nil
+	}
+	ip := ips[0]
+	switch r.strategy {
+	case lib.DNSFirst:
+		return ip
+	case lib.DNSRoundRobin:
+		r.rrm.Lock()
+		// NOTE: This index approach is not stable and might result in returning
+		// repeated or skipped IPs if the records change during a test run.
+		ip = ips[int(r.roundRobin[host])%len(ips)]
+		r.roundRobin[host]++
+		r.rrm.Unlock()
```

Comment:
I would argue this map should have a struct with the ... counter and a lock so you only lock one hosts roundRobin ... but given that I think you will need a lock on the outside map ... there might not be a performance improvement at all.

We should probably benchmark it still for the next iteration

Reply:
Yeah, that would be interesting to benchmark, but in general I'd like to avoid more mutexes :D 

I can add some benchmarks if you think they're required for this PR, otherwise I'd rather leave them for the proper resolver PR to avoid delaying the release further.

Reply:
No ... the only blocker for me is the fact we write the default values in the metadata.json ... you can ignore everything else until that is fixed :D

Reply:
Mostly everything else is what I have seen can be improved but arguably half of those will need benchmarking and should be left for another PR if at all